pkg/controllermanager/controller/shoot/shoot_maintenance_control.go

```diff
@@ -224,3 +221,101 @@ func hasMaintainNowAnnotation(shoot *gardenv1beta1.Shoot) bool {
 	operation, ok := shoot.Annotations[common.ShootOperation]
 	return ok && operation == common.ShootOperationMaintain
 }
+
+// MaintainMachineImage determines if a shoots machine image has to be maintained and in case returns the target image
+func MaintainMachineImage(shoot *gardenv1beta1.Shoot, cloudProfile *gardenv1beta1.CloudProfile, shootCurrentImage *gardenv1beta1.ShootMachineImage) (bool, *gardenv1beta1.ShootMachineImage, error) {
+	machineImagesFound, machineImageFromCloudProfile, err := helper.DetermineMachineImageForName(*cloudProfile, shootCurrentImage.Name)
+	if err != nil {
+		return false, nil, fmt.Errorf("failure while determining the machine image in the CloudProfile: %s", err.Error())
+	}
+	if !machineImagesFound {
+		return false, nil, fmt.Errorf("failure while determining the machine image in the CloudProfile: no machineImage with name '%s' (specified in shoot) could be found in the cloud cloudProfile '%s'", shootCurrentImage.Name, cloudProfile.Name)
+	}
+
+	machineImageExistsInCloudProfile, err := shootMachineImageExistsInCloudProfile(shoot, machineImageFromCloudProfile)
+	if err != nil {
+		return false, nil, fmt.Errorf("failure while checking if the shoot's machine image exists in the CloudProfile: %s", err.Error())
+	}
+
+	if !machineImageExistsInCloudProfile {
+		return updateToLatestMachineImageVersion(machineImageFromCloudProfile)
+	}
+
+	if *shoot.Spec.Maintenance.AutoUpdate.MachineImageVersion || ForceUpdateRequired(shootCurrentImage, machineImageFromCloudProfile) {
+		return canUpdateToLatestMachineImageVersion(shootCurrentImage, machineImageFromCloudProfile)
```

Comment:
Why don't you simply use `updateToLatestMachineImageVersion` here as well?
In `canUpdateToLatestMachineImageVersion` you are again validating whether the latest machine image version in the cloud profile is greater than the current shoot machine image version. What for? I think you can simply remove this code because it's not needed, correct? In this `if` condition either a force update is required because the expiration date of the currently used shoot machine image version expired or auto-update is desired. In both cases we always want to set the latest machine image version in the shoot spec. That's exactly what `updateToLatestMachineImageVersion` does.

```suggestion
		return updateToLatestMachineImageVersion(machineImageFromCloudProfile)
```

Reply:
I think I cannot remove the code because of the following scenario: 

AutoImageUpdate = true
```
gardenv1beta1.ShootMachineImage{
				Name:    "CoreOs",
				Version: "1.1.1",
			}}
```
& the Cloud Profile contains versions multiple versions lower than 1.1.1 and the version 1.1.1 as the highest version. Consequently the shoot is already on the highest version.

If we would adjust the coding as suggested, we would determine that we have to update to the latest version 1.1.1 (although we are already on that exact version). This would result in an unnecessary API Server call.
Also before executing the function canUpdateToLatestMachineImageVersion() we have not checked if there is a higher semantic version in the cloud profile.

I added another test that covers that scenario.

Reply:
> If we would adjust the coding as suggested, we would determine that we have to update to the latest version 1.1.1 (although we are already on that exact version). This would result in an unnecessary API Server call.

IMHO this doesn't matter compared to the code we are saving and the better readability. That's how it was before: When the machine image should be maintained, we simply update to the image that was referenced in the cloud profile. We don't have any performance issues whatsoever, so let's go ahead and remove this code.

> Also before executing the function canUpdateToLatestMachineImageVersion() we have not checked if there is a higher semantic version in the cloud profile.

Not needed, in the "worst-case" we make a no-op.